Extensions/fonts/freetype.lisp

```diff
@@ -552,3 +559,108 @@ or NIL if the current transformation is the identity transformation."
+;;; Font replacement code
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defgeneric text-style-fallback-fonts (text-style)
+  (:method ((text-style clim:standard-text-style))
+    nil))
+
+(defclass font-replacement-text-style (clim:standard-text-style)
+  ((fallback-fonts :initarg :replacement
+                   :reader font-replacement-text-style/replacement)))
+
+(defmethod text-style-fallback-fonts ((text-style font-replacement-text-style))
+  (font-replacement-text-style/replacement text-style))
+
+(defmethod clim:merge-text-styles ((style1 font-replacement-text-style) (style2 clim:standard-text-style))
+  (let ((merged-standard (call-next-method)))
+    (make-font-replacement-text-style (clim:text-style-family merged-standard)
+                                      (clim:text-style-face merged-standard)
+                                      (clim:text-style-size merged-standard)
+                                      (font-replacement-text-style/replacement style1))))
+
+(defmethod clim:merge-text-styles ((style1 clim:standard-text-style) (style2 font-replacement-text-style))
+  (let ((merged-standard (call-next-method)))
+    (make-font-replacement-text-style (clim:text-style-family merged-standard)
+                                      (clim:text-style-face merged-standard)
+                                      (clim:text-style-size merged-standard)
+                                      (font-replacement-text-style/replacement style2))))
+
+;;; The text styles needs to be cached because the implementation of
+;;; CLIM-CLX::LOOKUP-TEXT-STYLE-TO-X-FONT calls
+;;; CLIM:TEXT-STYLE-MAPPING in order to avoid doing a full font
+;;; lookup. The implementation of TEXT-STYLE-MAPPING ends up doing a
+;;; hash lookup on the text-style instance, which for CLOS instances
+;;; uses object identity.
+;;;
+;;; Having this cache effectively interns instances of
+;;; FONT-REPLACEMENT-TEXT-STYLE, allowing the TEXT-STYLE-MAPPING cache
+;;; to work properly.
+(defvar *text-styles* (make-hash-table :test 'equal))
+
+(defun make-font-replacement-text-style (family face size replacement)
+  (alexandria:ensure-gethash (list family face size replacement)
+                             *text-styles*
+                             (make-instance 'font-replacement-text-style
+                                            :text-family family
+                                            :text-face face
+                                            :text-size size
+                                            :replacement replacement)))
+
+(defun find-best-font (ch)
+  (let* ((match (mcclim-fontconfig:match-font `((:charset . (:charset ,ch))) '(:family :style) :kind :match-font)))
+    (let ((family (cdr (assoc :family match)))
+          (style (cdr (assoc :style match))))
+      (cond ((and family style)
+             (list family style))
+            (t
+             '(nil nil))))))
+
+(defun text-style-contains-p (port text-style ch)
```

Comment:
Text style contains what p? Again better name would be `text-style-contains-char-p`

Reply:
You are right. Fixed in ba0a66d81796e6dcc29c2693d4ea7ee33cdc3a5e